### PR TITLE
Implement Silver Mueller boundary conditions

### PR DESCRIPTION
**Pull request by RemiLehe**
The Silver-Mueller boundary is a simple absorbing boundary, that can serve as an alternative to the PML. 
Although it is not as efficient at absorbing waves, it does have much lower computational demand.

I also included an automated test, which is illustrate below: a laser is emitted towards the Silver-Mueller boundary, and is absorbed:
movie

The algorithm is described in this slide set, for instance: https://people.nscl.msu.edu/~lund/uspas/scs_2016/lec_adv/A1b_EM_Waves.pdf

(The above slides show the formula for `E`, but I actually implemented the ones for `B` in WarpX - which are symmetrical. I used `B` specifically, because - for a given boundary - the transverse B is cell-centered in the direction normal to the boundary, so the first value encountered outside of the physical domain is a `B` value - with the next `E` value being located further away from the boundary.)

Many thanks to @jlvay for explaining and clarifying some aspects of the Silver-Mueller boundaries off-line.

### Tests

- [x] propagation in `x`
- [ ] propagation in `y`
- [x] propagation in `z`

### CODE REVIEW

File: Examples/Tests/SilverMueller/inputs_2d_x [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Maximum number of time steps
```
Comment (RemiLehe) [inline Examples/Tests/SilverMueller/inputs_2d_x:1]: With this file (`inputs_2d_x`), the laser is emitted along the `x` axis.

File: Examples/Tests/SilverMueller/inputs_2d_z [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Maximum number of time steps
```
Comment (RemiLehe) [inline Examples/Tests/SilverMueller/inputs_2d_z:1]: In this file (`inputs_2d_z`), the laser is emitted along the `z` axis.

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        // Extract the tileboxes for which to loop
+        Box tbx  = mfi.tilebox(Bfield[0]->ixType().toIntVect());
+        Box tby  = mfi.tilebox(Bfield[1]->ixType().toIntVect());
+        Box tbz  = mfi.tilebox(Bfield[2]->ixType().toIntVect());
+
+        // We will modify the first (i.e. innermost) guard cell
+        // (if it is outside of the physical domain)
+        // Thus, the tileboxes here are grown by 1 guard cell
+        tbx.grow(1);
```
Comment (RemiLehe) [inline Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:74]: @WeiqunZhang Instead of doing this (growing the tiles themselves), could I modify the `for ( MFIter mfi(*Efield[0], false); mfi.isValid(); ++mfi ) {` to create tiles that perfectly divide the local box **extended by one guard cells**, without the tiles overlapping each other.

**Comment by ax3l** [comment]
I did a test with a quick ion-setup for a users and this looks great in a more complex scenario, too.

Ex_SM
Ex_PML

Ey_SM
Ey_PML

[inputs_2d.txt](https://github.com/ECP-WarpX/WarpX/files/6104416/inputs_2d.txt)

[Notebook.zip](https://github.com/ECP-WarpX/WarpX/files/6104444/Untitled.zip)

The runtime overhead is compared to no boundaries negligible :tada: and is ~20% faster than PMLs at the moment (even on a single Cori Haswell node with 64 MPI ranks, no OpenMP).

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    amrex::Real coef1_x = (1. - cdt_over_dx)/(1. + cdt_over_dx);
+    amrex::Real coef2_x = 2*cdt_over_dx/(1. + cdt_over_dx) / PhysConst::c;
+#ifdef WARPX_DIM_3D
+    amrex::Real const cdt_over_dy = PhysConst::c*dt*m_stencil_coefs_y[0];
+    amrex::Real coef1_y = (1. - cdt_over_dy)/(1. + cdt_over_dx);
+    amrex::Real coef2_y = 2*cdt_over_dy/(1. + cdt_over_dy) / PhysConst::c;
+#endif
+    amrex::Real const cdt_over_dz = PhysConst::c*dt*m_stencil_coefs_z[0];
+    amrex::Real coef1_z = (1. - cdt_over_dz)/(1. + cdt_over_dx);
+    amrex::Real coef2_z = 2*cdt_over_dz/(1. + cdt_over_dz) / PhysConst::c;
```
Comment (ax3l) [inline Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp]: If you like you can `const`-ify those. Let's also avoid the `int`s and potential `float`/`double` mixes:
```suggestion
    amrex::Real const coef1_x = (1._rt - cdt_over_dx)/(1._rt + cdt_over_dx);
    amrex::Real const coef2_x = 2._rt*cdt_over_dx/(1._rt + cdt_over_dx) / PhysConst::c;
#ifdef WARPX_DIM_3D
    amrex::Real const cdt_over_dy = PhysConst::c*dt*m_stencil_coefs_y[0];
    amrex::Real const coef1_y = (1._rt - cdt_over_dy)/(1._rt + cdt_over_dx);
    amrex::Real const coef2_y = 2._rt*cdt_over_dy/(1._rt + cdt_over_dy) / PhysConst::c;
#endif
    amrex::Real const cdt_over_dz = PhysConst::c*dt*m_stencil_coefs_z[0];
    amrex::Real const coef1_z = (1._rt - cdt_over_dz)/(1._rt + cdt_over_dx);
    amrex::Real const coef2_z = 2._rt*cdt_over_dz/(1._rt + cdt_over_dz) / PhysConst::c;
```

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                // At the -x boundary (innermost guard cell)
+                if ( i==domain_box.smallEnd(0)-1 )
+                    Bz(i,j,k) = coef1_x * Bz(i,j,k) - coef2_x * Ey(i+1,j,k);
+#ifdef WARPX_DIM_3D
+                // At the +y boundary (innermost guard cell)
+                if ( j==domain_box.bigEnd(1)+1 )
+                    Bz(i,j,k) = coef1_x * Bz(i,j,k) - coef2_y * Ex(i,j,k);
+                // At the -y boundary (innermost guard cell)
+                if ( j==domain_box.smallEnd(1)-1 )
+                    Bz(i,j,k) = coef1_x * Bz(i,j,k) + coef2_y * Ex(i,j+1,k);
```
Comment (ax3l) [inline Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp]: Looks like a 3D bug:
```suggestion
                    Bz(i,j,k) = coef1_y * Bz(i,j,k) - coef2_y * Ex(i,j,k);
                // At the -y boundary (innermost guard cell)
                if ( j==domain_box.smallEnd(1)-1 )
                    Bz(i,j,k) = coef1_y * Bz(i,j,k) + coef2_y * Ex(i,j+1,k);
```

Comment (ax3l) [reply Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp]: Following Murphy's law, this hid in the only direction of propagation that we did not test.

**Comment by ax3l** [comment]
To do for this PR as discussed offline: We'll take a look at this comment https://github.com/ECP-WarpX/WarpX/pull/1766#discussion_r589700523 and then add another test for the `y` direction.

File: Source/WarpX.cpp [hunk 6]
```diff
@@ -554,6 +554,10 @@ WarpX::ReadParameters ()
         }
 
         pp.query("do_pml", do_pml);
+        pp.query("do_silver_mueller", do_silver_mueller);
+        if ( (do_pml==1)&&(do_silver_mueller==1) ) {
+            amrex::Abort("PML and Silver-Mueller boundary conditions cannot be activated at the same time.");
```
Comment (RevathiJambunathan) [inline Source/WarpX.cpp:559]: do we also need an abort for turning on Silver-Mueller with num_levels>1

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        tbz.grow(1);
+
+        // Apply Boundary condition to Bx
+        amrex::ParallelFor(tbx,
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+
+#ifdef WARPX_DIM_3D
+                // At the +y boundary (innermost guard cell)
+                if ( j==domain_box.bigEnd(1)+1 )
+                    Bx(i,j,k) = coef1_y * Bx(i,j,k) + coef2_y * Ez(i,j,k);
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:87]: The comment earlier says update Efield at the boundary, but, here the Bfields are getting updated.

**Review by RevathiJambunathan — CHANGES_REQUESTED** [review]
Sorry! I just read the description

**Review by RevathiJambunathan — COMMENTED** [review]
Thank you for this PR!

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp (hunk 7 above)
Comment (RevathiJambunathan) [reply Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:87]: oh never mind. I just read the PR description!!

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+#include "Utils/WarpXConst.H"
+#include <AMReX_Gpu.H>
+
+using namespace amrex;
+
+/**
+ * \brief Update the E field at the boundary, using the Silver-Mueller condition
+ */
+void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Efield,
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:19]: would Efield have a const before it then?

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+#include "Utils/WarpXAlgorithmSelection.H"
+#include "FiniteDifferenceSolver.H"
+#include "Utils/WarpXConst.H"
+#include <AMReX_Gpu.H>
+
+using namespace amrex;
+
+/**
+ * \brief Update the E field at the boundary, using the Silver-Mueller condition
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp]:
```suggestion
 * \brief Update the B field at the boundary, using the Silver-Mueller condition
```

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp (hunk 8 above)
Comment (RemiLehe) [reply Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:19]: Good point! Actually I removed the `const` for `B` now, as it was confusing.

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp (hunk 3 above)
Comment (WeiqunZhang) [reply Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:74]: Yes, you can do that.  You can also skip the boxes that do not touch the domain and merge the three kernels into one.  Here is a patch (though not tested).  (Github does not appear to support attaching a patch file. ?)

```diff
diff --git a/Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp b/Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp
index 6636a83..6f2f54c 100644
--- a/Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp
+++ b/Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp
@@ -50,10 +50,11 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
 #ifdef AMREX_USE_OMP
 #pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
 #endif
-    // tiling is usually set by TilingIfNotGPU()
-    // but here, we set it to false because of potential race condition,
-    // since we grow the tiles by one guard cell after creating them.
-    for ( MFIter mfi(*Efield[0], false); mfi.isValid(); ++mfi ) {
+    for ( MFIter mfi(*Efield[0], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+
+        if (domain_box.strictly_contains(amrex::enclosedCells(mfi.validbox()))) {
+            continue;
+        }
 
         // Extract field data for this grid/tile
         Array4<Real> const& Ex = Efield[0]->array(mfi);
@@ -64,19 +65,12 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
         Array4<Real> const& Bz = Bfield[2]->array(mfi);
 
         // Extract the tileboxes for which to loop
-        Box tbx  = mfi.tilebox(Bfield[0]->ixType().toIntVect());
-        Box tby  = mfi.tilebox(Bfield[1]->ixType().toIntVect());
-        Box tbz  = mfi.tilebox(Bfield[2]->ixType().toIntVect());
-
-        // We will modify the first (i.e. innermost) guard cell
-        // (if it is outside of the physical domain)
-        // Thus, the tileboxes here are grown by 1 guard cell
-        tbx.grow(1);
-        tby.grow(1);
-        tbz.grow(1);
-
-        // Apply Boundary condition to Bx
-        amrex::ParallelFor(tbx,
+        const Box& tbx = mfi.tilebox(Bfield[0]->ixType().toIntVect(), IntVect(1));
+        const Box& tby = mfi.tilebox(Bfield[1]->ixType().toIntVect(), IntVect(1));
+        const Box& tbz = mfi.tilebox(Bfield[2]->ixType().toIntVect(), IntVect(1));
+
+        amrex::ParallelFor(tbx, tby, tbz,
+            // Apply Boundary condition to Bx
             [=] AMREX_GPU_DEVICE (int i, int j, int k){
 
 #ifdef WARPX_DIM_3D
@@ -100,11 +94,8 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
                 if ( j==domain_box.smallEnd(1)-1 )
                     Bx(i,j,k) = coef1_z * Bx(i,j,k) + coef2_z * Ey(i,j+1,k);
 #endif
-            }
-        );
-
-        // Apply Boundary condition to By
-        amrex::ParallelFor(tby,
+            },
+            // Apply Boundary condition to By
             [=] AMREX_GPU_DEVICE (int i, int j, int k){
 
                 // At the +x boundary (innermost guard cell)
@@ -128,11 +119,8 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
                 if ( j==domain_box.smallEnd(1)-1 )
                     By(i,j,k) = coef1_z * By(i,j,k) - coef2_z * Ex(i,j+1,k);
 #endif
-            }
-        );
-
-        // Apply Boundary condition to Bz
-        amrex::ParallelFor(tbz,
+            },
+            // Apply Boundary condition to Bz
             [=] AMREX_GPU_DEVICE (int i, int j, int k){
 
                 // At the +x boundary (innermost guard cell)
```

**Review by RevathiJambunathan — APPROVED** [review]

File: Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp (hunk 8 above)
Comment (ax3l) [reply Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:19]: I think the `Efield` is only read, so array, pointer and MultiFab are `const`.

The `Bfield` is manipulated, so only the array and pointer are const, not the MultiFab

**Review by ax3l — COMMENTED** [review]
Please don't forget to add a test for y absorption (`silver_mueller_2d_y`) :)

**Review by ax3l — APPROVED** [review]
Will merge it for now since it's needed for a paper and the 3D will be added as a follow-up.

File: Source/WarpX.cpp (hunk 6 above)
Comment (ax3l) [reply Source/WarpX.cpp:559]: yes, we don't do this generally yet in the code when the mesh refinement patch touches the global boundary. We need to fix this for all boundary conditions in a separate issue.
